### ISSUE TITLE
Create `TraitPass<T>` and `InterfacePass<T>` to mirror `OperationPass<T>`

### ISSUE THREAD

**Issue by stellaraccident** [llvm/llvm-project#52916]
The `OperationPass<T>` template/base class is useful for a few reasons for defining passes which have to operate on a specific operation:

* It internally checks that the operation is of the given type, providing a good error if not.
* As documentation, it signals the constraint to callers so that it is obvious how to setup a nested pass manager.
* When using implicit nesting, it will nest automatically.

The first two characteristics would also be useful for other constrained passes which must operate on operations with a given trait or interface. Propose adding `TraitPass<T>` and `InterfacePass<T>` to capture this. In-tree, the LinalgDetensorize pass would benefit from the first.

**Comment by River707** [comment]
I have a few ideas here for a more general feature that should support this, will play around with it this week.

**Comment by stellaraccident** [comment]
Thanks, River for implementing this. While cleaning up our code to use this, I realized that this does indeed help us be more type-safe with respect to implementing FunctionLike-constrained passes, but it is still pretty awkward to schedule them. For example, we have a utility `FunctionLikeNest` which schedules a pass constructor against all of our known top-level implementors of `FunctionLike`.

An example of doing so:

```
using FunctionLikeNest = MultiOpNest<FuncOp, IREE::Util::InitializerOp>;

FunctionLikeNest(pipeline)
    .addPass(createInterchangeGenericOpsPass)
    .addPass(mlir::createCanonicalizerPass)
    .addPass(mlir::createCSEPass);
```

I had hoped to someday have a better way of doing this. Was wondering if you had any suggestions?

**Comment by River707** [comment]
Yes, I have a few ideas here but haven't had the time to realize them (the FuncOp move has been taking up a lot). It's on my list, I may have something to share in the next few weeks.

**Comment by stellaraccident** [comment]
Thanks and no worries. InterfacePass makes a lot of our code more documenting and safer. Thanks for the incremental step.
